bots/heroes/dr_main.lua

```diff
@@ -0,0 +1,340 @@
+
+	local nlev = unitSelf:GetLevel()
+	local nlevpts = unitSelf:GetAbilityPointsAvailable()
+	for i = nlev, nlev+nlevpts do
+		unitSelf:GetAbility( object.tSkills[i] ):LevelUp()
+	end
+end
+
+-- @param: eventdata
+-- @return: none
+function object:oncombateventOverride(EventData)
+	local nBonus = 0
+
+	if EventData.Type == "Ability" then
+		if EventData.InflictorName == "Ability_DoctorRepulsor4" then
+			local lvl = core.unitSelf:GetLevel()
+			if lvl >= 16 then
+				nBonus = nBonus + object.nSpeedUse + 10 * lvl / 10
+			elseif lvl >= 11 then
+				nBonus = nBonus + object.nSpeedUse + 5 * lvl / 10
+			else
+				nBonus = nBonus + object.nSpeedUse + (lvl / 3) * 2
+			end
+		elseif EventData.InflictorName == "Ability_DoctorRepulsor1" then
+			nBonus = nBonus + object.nContrpUse
+		elseif EventData.InflictorName == "Ability_DoctorRepulsor2" then
+			nBonus = nBonus + object.nOpposingUse
+		else
+			nBonus = nBonus + object.nContrpUse
+		end
+	elseif EventData.Type == "Item" and EventData.SourceUnit == core.unitSelf:GetUniqueID() then
+		local itemMorph, itemSilence = core.GetItem("Item_Morph"), core.GetItem("Item_Silence")
+		if itemMorph ~= nil and EventData.InflictorName == itemMorph:GetName() then
+			nBonus = nBonus + object.nMorphUse
+		elseif itemSilence ~= nil and EventData.InflictorName == itemSilence:GetName() then
+			nBonus = nBonus + object.nHFlowerUse
+		end
+	end
+
+	if nBonus > 0 then
+		core.DecayBonus(self)
+		core.nHarassBonus = core.nHarassBonus + nBonus
+	end
+end
+object.oncombateventOld = object.oncombatevent
+object.oncombatevent     = object.oncombateventOverride
+
+-- @param: iunitentity hero
+-- @return: number
+local function CustomHarassUtilityFnOverride(hero)
+	local nValue = 0
+
+	if skills.abilQ:CanActivate() then
+		nValue = nValue + object.nContrpUp
+	end
+
+	if skills.abilW:CanActivate() then
+		nValue = nValue + object.nOpposingUp
+	end
+
+	if skills.abilR:CanActivate() then
+		nValue = nValue + object.nSpeedUp
+	end
+
+	local itemMorph = core.GetItem("Item_Morph")
+	if itemMorph and itemMorph:CanActivate() then
+		nValue = nValue + object.nMorphUp
+	end
+
+	local itemSilence = core.GetItem("Item_Silence")
+	if itemSilence and itemSilence:CanActivate() then
+		nValue = nValue + object.nHFlowerUp
+	end
+
+	return nValue
+end
+behaviorLib.CustomHarassUtility = CustomHarassUtilityFnOverride  
+
+-- @param botBrain: CBotBrain
+-- @return: none
+local function CustomRetreatExecuteFnOverride(botBrain)
+	-- If we're low HP, run for our life!
+	local unitSelf = core.unitSelf
+	if unitSelf and unitSelf:GetHealth() > 0 and unitSelf:GetHealth() < 400
+		and skills.abilR:CanActivate() and unitSelf:GetMana() > 230
+	then
+		-- stolen from RallyTest and adjusted a bit to fit
+		local vecOrigin = unitSelf:GetPosition()
+		local vecDirection = Vector3.Create(1, 0)
+		vecDirection = core.RotateVec2D(vecDirection, unitSelf:GetTeam() == 2 and 90 or -90)
+		core.OrderAbilityPosition(self, skills.abilR, vecOrigin + vecDirection * 500)
```

Comment:
self is nil here, should be botBrain.
Also could use behaviorLib.GetSafeBlinkPosition(vecDesiredPosition, nRange) to find a position that is not 'stuck'
position would be well - core.allyWell:GetPosition() and range what ever you like.
Ill merge after that.